### PR TITLE
Add generate() functionality to TF 2.0

### PR DESCRIPTION
**Pull request by patrickvonplaten**
I added the `_generate_no_beam_search` functionality for TF 2.0. It works for the following models:
'gpt2', 'openai', 'xlnet', 'xlm', 'ctrl'. Only for the model 'transfo-xl' it doesn't work because the 
lm_head is not implemented yet in TF 2.0 (added an issue here: #3064 ). 

Also I checked whether the pytorch 'distilgpt2' and TF 2.0 'distilgpt2' generate the same output (added one Integration test for this). Will add other integration tests in a future PR.

Setting only certain indices to values is much less straight forward in TF 2.0 than in pytorch, which is why I added more code for the TF 2.0 version. 

Would be very happy about some feedback @LysandreJik @thomwolf 

EDIT: There was also a bug in TFCTRL where tf.concat uses a pytorch argument 'dim' instead of 'axis'

## TODO
- [x] Discuss how to change the test_torch_tf_conversion.py() @LysandreJik @thomwolf 

## Future PR:
- [ ] Adapt all LMHead Integration Tests to greedy generate to be able to compare PT & TF
- [ ] Add generate() to TFTransfoXL (see Issue #3064 )

### CODE REVIEW

File: tests/test_modeling_gpt2.py [hunk 1]
```diff
@@ -310,33 +310,33 @@ def test_lm_generate_gpt2(self):
...
-            3621,
-            284,
-        ]  # The dog is cute though he can sometimes just walk in the park. It is not very nice to
-        torch.manual_seed(0)
+            2635,
+        ]  # The president of the United States, and the president of the United Kingdom, have been in the White
 
         output_ids = model.generate(
             input_ids,
+            do_sample=False,
```
Comment (patrickvonplaten) [inline tests/test_modeling_gpt2.py]: @thomwolf @LysandreJik For the LM Head integration tests I want to change to greedy search for model.generate() to be able to compare tf and pt (setting the same random seed in TF and pytorch does not produce same random results)

File: tests/test_modeling_tf_common.py [hunk 2]
```diff
@@ -354,6 +355,46 @@ def test_inputs_embeds(self):
...
+        for model_class in self.all_generative_model_classes:
+            model = model_class(config)
+
+            if config.bos_token_id is None:
+                with self.assertRaises(AssertionError):
+                    model.generate(max_length=5)
+                # batch_size = 1
+                self._check_generated_tokens(model.generate(input_ids))
+                # batch_size = 1, num_beams > 1
+            #                self._check_generated_tokens(model.generate(input_ids, num_beams=3))  #TODO: beam_search will be implemented at a later stage
```
Comment (patrickvonplaten) [inline tests/test_modeling_tf_common.py]: Re-Add the beam_search > 1 tests later when beam_search is figured out.

File: tests/test_modeling_tf_gpt2.py [hunk 3]
```diff
@@ -234,3 +250,48 @@ def test_model_from_pretrained(self):
...
+            7526,
+            11,
+            423,
+            587,
+            287,
+            262,
+            2635,
+        ]  # The president of the United States, and the president of the United Kingdom, have been in the White
+
+        output_ids = model.generate(
```
Comment (patrickvonplaten) [inline tests/test_modeling_tf_gpt2.py]: produces same results in TF as in PT for greedy search

File: tests/test_torch_tf_conversion.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+# coding=utf-8
```
Comment (patrickvonplaten) [inline tests/test_torch_tf_conversion.py]: @thomwolf, @LysandreJik I re-wrote the top_k_top_p_filtering function in tensorflow and since it's quite an important function I feel like we should at least test that TF and PT produce the same results. Not quite sure about the testing format. This file is a version for how I would test it. What do you think @thomwolf , @LysandreJik ?

File: src/transformers/modeling_tf_utils.py [hunk 5]
```diff
@@ -384,6 +384,431 @@ def from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs):
...
+        top_p,
+        repetition_penalty,
+        pad_token_id,
+        eos_token_ids,
+        batch_size,
+        length_penalty,
+        num_beams,
+        vocab_size,
+    ):
+        pass
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py]: will be added when PT generate_beam_search works cleanely!

File: src/transformers/modeling_ctrl.py [hunk 6]
```diff
@@ -454,14 +454,12 @@ def __init__(self, config):
     def get_output_embeddings(self):
         return self.lm_head
 
-    def prepare_inputs_for_generation(self, input_ids, **kwargs):
+    def prepare_inputs_for_generation(self, input_ids, past, **kwargs):
```
Comment (LysandreJik) [inline src/transformers/modeling_ctrl.py]: Cleaner!

File: src/transformers/modeling_tf_ctrl.py [hunk 7]
```diff
@@ -105,8 +105,8 @@ def call(self, inputs, training=False):
         v = self.split_into_heads(v, batch_size)
         if layer_past is not None:
             past_key, past_value = tf.unstack(layer_past, axis=1)
-            k = tf.concat((past_key, k), dim=-2)
-            v = tf.concat((past_value, v), dim=-2)
+            k = tf.concat((past_key, k), axis=-2)
+            v = tf.concat((past_value, v), axis=-2)
```
Comment (LysandreJik) [inline src/transformers/modeling_tf_ctrl.py]: nice catch!

File: tests/test_torch_tf_conversion.py (hunk 4 above)
Comment (LysandreJik) [reply tests/test_torch_tf_conversion.py]: Sure, testing that they both output the same results is cool. Wouldn't we want to test that they output a specific output that we know for sure to be the correct top_p/top_k, to test that they both actually behave as we want? Would seem more robust to me.

**Review by LysandreJik — APPROVED** [review]
This is cool, good job on the TensorFlow implementation! Regarding the use of the `past` and `mems`, I don't think they're actually implemented in the models though?

File: tests/test_torch_tf_conversion.py (hunk 4 above)
Comment (patrickvonplaten) [reply tests/test_torch_tf_conversion.py]: changed the tests to hardcoded individual integration tests!

File: tests/test_modeling_common.py [hunk 8]
```diff
@@ -695,3 +705,96 @@ def test_model_from_pretrained(self):
...
             self.assertEqual(model.config, config)
+
+
+@require_torch
+class UtilsFunctionsTest(unittest.TestCase):
+
+    # tests whether the top_k_top_p function behaves as expected
+    def test_top_k_top_p_filtering(self):
+        if is_torch_available():
+            sys.path.insert(1, "../src")
```
Comment (patrickvonplaten) [inline tests/test_modeling_common.py]: seems that the path '../src' has to be inserted to the PYTHONPATH before import since 
`from transformers import top_k_top_p_filtering` does not work.

**Comment by patrickvonplaten** [comment]
> This is cool, good job on the TensorFlow implementation! Regarding the use of the `past` and `mems`, I don't think they're actually implemented in the models though?

I think the tf models also have the `past` and `mems` functionality implemented (when looking into the tf modeling files the `past` and `mems` variables are used in the code.

File: src/transformers/modeling_tf_utils.py [hunk 9]
```diff
@@ -384,6 +384,432 @@ def from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs):
...
+
+
+def tf_top_k_top_p_filtering(logits, top_k=0, top_p=1.0, filter_value=-float("Inf"), min_tokens_to_keep=1):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py]: I implemented this function from its torch version `top_k_top_p_filtering(...)` in https://github.com/huggingface/transformers/blob/0e56b37e805279ecb61670159fa8c71487214e0a/src/transformers/modeling_utils.py#L1127 

Some operations like setting tensor indices to certain values or scattering values to indices seemed much harder in TF 2.0 than in torch. That's why I added the two functions `scatter_values_on_batch_indices` and `set_tensor_by_indices_to_value`. Would be happy about some feedback concerning whether the "correct" tf functions were used for the given function @jplu

**Review by thomwolf — APPROVED** [review]

File: src/transformers/modeling_tf_utils.py (hunk 9 above)
Comment (jplu) [reply src/transformers/modeling_tf_utils.py]: For what I'm seeing of what has been done, everything seems ok from a TF point of view :) The only one thing is I always prefer to use the tf function when they are available for example line 758 use `tf.shape(logits)` instead of `logits.shape` but this is just a personal usage, it is really good as it is :+1:

**Comment by patrickvonplaten** [comment]
Good to merge for me! 
Changed all `<tf_tensor>.shape` to `shape_list(<tf_tensor>)` to make function compatible with both eagermode and no eagermode after discussion with @thomwolf  and @jplu .

Will add additional Integration tests (so far only `tf_gpt2`) for other LMHead models.

**Review by LysandreJik — APPROVED** [review]
LGTM
